test/openssl/test_pkey.rb

```diff
@@ -82,6 +82,13 @@ def test_hmac_sign_verify
   end
 
   def test_ed25519
+    # Ed25519 is not approved in OpenSSL 3.1+ FIPS code.
+    # See OpenSSL providers/fips/fipsprov.c PROV_NAMES_ED25519 entries
+    # with FIPS_UNAPPROVED_PROPERTIES in OpenSSL 3.1+.
+    if OpenSSL.fips_mode && openssl?(3, 1, 0, 0)
+      omit "Ed25519 is not approved in OpenSSL 3.1+ FIPS code"
+    end
```

Comment:
Is this a change in OpenSSL 3.1's fips provider? I assumed Ed25519 would be unavailable in all FIPS-compliant systems including older versions of OpenSSL.

I wonder if we can use the `omit_on_fips` added in the 4th commit in this case, too.

Reply:
The situation is a bit complicated.

OpenSSL banned the Ed25519 by policy. The policy means they documented in [their security policy documents for openssl 3.0.0 and 3.0.8](https://www.openssl.org/source/) that you can download from the link.

> The module implements the following non-Approved algorithms:
> 
> Ed448 - Digital Signature Generation
> Ed25519 - Digital Signature Generation
> Table 8 – Non-Approved Algorithms
> 
> These algorithms shall not be used when operating in the FIPS Approved mode of operation. Use of the non-Approved algorithms
listed in the table above will place the module in the non-Approved mode of operation.

However, their OpenSSL 3.0 code technically allows Ed25519 in FIPS, raising the error.
https://github.com/openssl/openssl/blob/50f3f93753c9b29ae37895fa1fa6280c1066ba67/providers/fips/fipsprov.c#L394

Below are related comments about the topic.

https://github.com/openssl/openssl/issues/20758#issuecomment-1638063764
https://github.com/openssl/openssl/issues/20758#issuecomment-1638521876
https://github.com/openssl/openssl/issues/20758#issuecomment-1639658102
> The technical committee discussed this at length and decided that the paperwork change banning these algorithms was good enough and that a patch to also "fix" the code wasn't required. This approach is fine under the FIPS 140-2 standard that we validated even though it places extra burden on the users of the validation.

> I wonder if we can use the omit_on_fips added in the 4th commit in this case, too.

All right. I will try to use the `omit_on_fips` in the case.

Reply:
Now I am using the `omit_on_fips` method in the `test_ed25519` too.

Reply:
Interesting. I didn't know it was possible to implement a non-approved algorithm while being validated, so I thought it was not implemented in 3.0 either.

However, from the discussions, it seems that the inclusion of Ed25519 in OpenSSL 3.0's fips provider was not intentional and users are prohibited to use this feature even if it may appear to work. I understand they chose not to fix the code because that would have required them to redo all validation.

IMHO, `test_ed25519` should be skipped regardless of the OpenSSL version if Ed25519 must not be used anyway. I don't want to fix this test case again if a future OpenSSL 3.0.x release starts to properly prohibit Ed25519 with the fips provider (though it may be unlikely considering that OpenSSL 3.1 is already out).

Reply:
To clarify, I don't think we should be testing OpenSSL bugs.

I think there may be cases where we still want to skip based on the OpenSSL version number, for example, due to the difference between FIPS 140-2 (OpenSSL 3.0) and FIPS 140-3 (OpenSSL 3.1), but this was different.

Reply:
> I don't want to fix this test case again if a future OpenSSL 3.0.x release starts to properly prohibit Ed25519 with the fips provider (though it may be unlikely considering that OpenSSL 3.1 is already out).

> To clarify, I don't think we should be testing OpenSSL bugs.

OK. I can understand your concerns above. I removed the `openssl?(3, 1, 0, 0)` in the `test_ed25519`.